File: sol/solution/0001-0100/0031. Next Permutation/Solution.rb
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def next_permutation(nums)
    i = nums.length - 2
    i -= 1 while i >= 0 && nums[i] >= nums[i + 1]
    if i >= 0
        j = nums.length - 1
        j -= 1 while nums[j] <= nums[i]
        nums[i], nums[j] = nums[j], nums[i]
    end
    nums[i+1..-1] = nums[i+1..-1].reverse
end

File: sol/solution/0001-0100/0031. Next Permutation/Solution.go
func nextPermutation(nums []int) {
    i := len(nums) - 2
    for i >= 0 && nums[i] >= nums[i+1] {
        i--
    }
    if i >= 0 {
        j := len(nums) - 1
        for nums[j] <= nums[i] {
            j--
        }
        nums[i], nums[j] = nums[j], nums[i]
    }
    reverse(nums[i+1:])
}
func reverse(nums []int) {
    l, r := 0, len(nums)-1
    for l < r {
        nums[l], nums[r] = nums[r], nums[l]
        l++
        r--
    }
}